### DISCUSSION TITLE
User is Logged in check like firebase.

### DISCUSSION THREAD

**Discussion by sabikrahat** [pocketbase/pocketbase#1887] [Q&A] [answered]
Hello!
In firebase there is a way to check user isLogged in by `FirebaseAuth.instance.currentUser != null`. Here in pocketbase after `authWithPassword` we can get `token and model` from `pb.authStore`. But whenever we again relaunch the app, this `pb.authStore` always return `token == null`. Is there any way to store this like firebase without storing in local database manually?

**Comment by ganigeorgiev** [comment, marked as answer]
Do you mean the Dart SDK?

If Yes, then you can also check https://github.com/pocketbase/dart-sdk/issues/13.
The `AuthStore` storage is not persistent at the moment out of the box because there are too many different ways to setup it (encrypted or plain localstore, sharedpreferences, key-value databases, etc.).

You can create your own AuthStore by extending the [base one](https://pub.dev/documentation/pocketbase/latest/pocketbase/AuthStore-class.html) as shown in https://github.com/pocketbase/dart-sdk#authstore (generally you have to overwrite only the `save(token, model)` and `clear()` methods).

**Reply by sabikrahat** [reply]
Okay.... can you give a short example of CustomAuthStore to overriding save and clear function using any local db like sharedpreference or isar or hive?

**Reply by ganigeorgiev** [reply]
Please read https://github.com/pocketbase/dart-sdk/issues/13#issuecomment-1254558912.

There are too many options for Flutter and it is better to choose the right storage mechanism based on what you feel more comfortable working with.

If you need a reference, here is a quick exampe with shared_preferences (I haven't tested it):

```dart
class CustomAuthStore extends AuthStore {
  final SharedPreferences prefs;
  final String key;

  CustomAuthStore(this.prefs, {this.key = "pb_auth"}) {
    final String? raw = prefs.getString(key);

    if (raw != null && raw.isNotEmpty) {
      final decoded = jsonDecode(raw);
      final token = (decoded as Map<String, dynamic>)["token"] as String? ?? "";
      final model =
          RecordModel.fromJson(decoded["model"] as Map<String, dynamic>? ?? {});

      save(token, model);
    }
  }

  @override
  void save(
    String newToken,
    dynamic /* RecordModel|AdminModel|null */ newModel,
  ) {
    super.save(newToken, newModel);

    final encoded =
        jsonEncode(<String, dynamic>{"token": newToken, "model": newModel});
    prefs.setString(key, encoded);
  }

  @override
  void clear() {
    super.clear();

    prefs.remove(key);
  }
}

....

final prefs = await SharedPreferences.getInstance();

final pb = PocketBase("http://127.0.0.1:8090", authStore: CustomAuthStore(prefs));
```

Once again, the above is just a rudimentary example and can be implemented in a lot of different ways.

Alternatively, as mentioned in https://github.com/pocketbase/dart-sdk/issues/13#issuecomment-1254558912, you can also just use the default AuthStore and sync it with your custom one using the `onChange` stream, something like:

```dart
final pb = PocketBase("http://127.0.0.1:8090");

final prefs = await SharedPreferences.getInstance();

// load the initial pb.authStore state from the SharedPreferences store
final String? raw = prefs.getString("pb_auth");
if (raw != null && raw.isNotEmpty) {
  final decoded = jsonDecode(raw);
  final token = (decoded as Map<String, dynamic>)["token"] as String? ?? "";
  final model =
      RecordModel.fromJson(decoded["model"] as Map<String, dynamic>? ?? {});

  pb.authStore.save(token, model);
}

// sync the SharedPreferences store with the auth change
pb.authStore.onChange.listen((e) {
  final encoded = jsonEncode(<String, dynamic>{
    "token": e.token,
    "model": e.model,
  });
  prefs.setString("pb_auth", encoded);
});
```

**Reply by sabikrahat** [reply]
Thanks a lot....❣️

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here, Dart SDK v0.13.0 comes with `AsyncAuthStore` helper that could be used with any external async persistent layer (shared preferences, hive, local file, etc.):_

```dart
final prefs = await SharedPreferences.getInstance();

final store = AsyncAuthStore(
 save:    (String data) async => prefs.setString('pb_auth', data),
 initial: prefs.getString('pb_auth'),
);

final pb = PocketBase('http://example.com', authStore: store);
```
